pkg/cosign/fetch.go

```diff
@@ -138,20 +138,42 @@ func FetchAttestationsForReference(ctx context.Context, ref name.Reference, opts
 		return nil, fmt.Errorf("no attestations associated with %s", ref)
 	}
 
-	attestations := make([]AttestationPayload, len(l))
+	attestations := make([]AttestationPayload, 0, len(l))
 	var g errgroup.Group
 	g.SetLimit(runtime.NumCPU())
+
 	for i, att := range l {
-		i, att := i, att
+		if predicateURI != "" {
+			anns, err := att.Annotations()
+			if err != nil {
+				return nil, err
+			}
+			pt, ok := anns["predicateType"]
+			// Skip attestation if predicateType annotation is not present or predicateType annotation does not match supplied predicate type
+			if !ok || pt != predicateURI {
+				continue
+			}
+		}
+		_, att := i, att
+		var a AttestationPayload
 		g.Go(func() error {
 			attestPayload, _ := att.Payload()
-			return json.Unmarshal(attestPayload, &attestations[i])
+			err := json.Unmarshal(attestPayload, &a)
+			if err != nil {
+				return err
+			}
+			attestations = append(attestations, a)
+			return nil
 		})
 	}
 	if err := g.Wait(); err != nil {
 		return nil, err
 	}
 
+	if len(attestations) == 0 && predicateURI != "" {
+		return nil, fmt.Errorf("no layers of predicate type %s found", predicateURI)
```

Comment:
you might want to use the following error message: `no attestations with predicate type '%s' found`